### PR TITLE
Support serializing core astropy classes with YAML

### PR DESCRIPTION
**Pull request by taldcroft**
This adds support for safe YAML serializing  (dump and load) of:
- np.ndarray
- Quantity, Angle, Latitude, Longitude
- Time, TimeDelta
- EarthLocation
- SkyCoord

The current implementation does *not* allow for arbitrary subclasses of these objects in order to ensure that the load process is "safe" and only uses trusted constructors.

This PR is on the path to allowing mixin columns in ECSV that fully round-trip.  It also relates somewhat to #5471.

@mhvk @cdeil @astrofrog

### CODE REVIEW

**Comment by taldcroft** [comment]
FYI, here is a somewhat complicated example:
```
In [7]:     c = SkyCoord([[1,2],[3,4]], [[5,6], [7,8]],
   ...:                  unit='deg', frame=frame,
   ...:                  obstime=Time.now(),
   ...:                  location=EarthLocation(1000, 2000, 3000, unit=u.km))

In [8]: print(dump(c))
!astropy.coordinates.sky_coordinate.SkyCoord
alt: !astropy.units.Quantity
  class: Latitude
  unit: &id001 !astropy.units.Unit {name: deg}
  value: !numpy.ndarray
    __ndarray__: !!binary |
      QUFBQUFBQUFGRUFBQUFBQUFBQVlRQUFBQUFBQUFCeEFBQUFBQUFBQUlFQT0=
    dtype: float64
    shape: !!python/tuple [2, 2]
az: !astropy.units.Quantity
  class: Longitude
  unit: *id001
  value: !numpy.ndarray
    __ndarray__: !!binary |
      QUFBQUFBQUE4RDhBQUFBQUFBQUFRQUFBQUFBQUFBaEFBQUFBQUFBQUVFQT0=
    dtype: float64
    shape: !!python/tuple [2, 2]
frame: altaz
location: !astropy.coordinates.earth.EarthLocation
  ellipsoid: WGS84
  x: !astropy.units.Quantity
    class: Quantity
    unit: &id002 !astropy.units.Unit {name: km}
    value: 1000.0
  y: !astropy.units.Quantity
    class: Quantity
    unit: *id002
    value: 2000.0
  z: !astropy.units.Quantity
    class: Quantity
    unit: *id002
    value: 3000.0
obstime: !astropy.time.Time {format: datetime, in_subfmt: '*', jd1: 2457711.5, jd2: 0.7801903949421296,
  out_subfmt: '*', precision: 3, scale: utc}
obswl: !astropy.units.Quantity
  class: Quantity
  unit: !astropy.units.Unit {name: micron}
  value: 1.0
pressure: !astropy.units.Quantity
  class: Quantity
  unit: !astropy.units.Unit {name: hPa}
  value: 0.0
relative_humidity: 0
representation: spherical
temperature: !astropy.units.Quantity
  class: Quantity
  unit: !astropy.units.Unit {name: deg_C}
  value: 0.0
```

**Comment by mhvk** [comment]
I like where this is going! But I'll need to think more to comment very sensibly. For now,
1. For `SkyCoord`, it turns out to be very handy that it has a list of things that are required to reproduce it faithfully. Perhaps we should have the same for other classes? E.g., in `Time._apply`, the list you have here is also used.
2. Slightly relatedly, to what extent could we combine the information on how to serialise the objects with what pickle does? In the end, it needs the same information. Similarly, any effort to serialise in JSON would need the same information as well. Could we rely more on `self.__dict__`?
3. There is something to be said for letting classes provide their own interface; this could perhaps be made part of `info`?

p.s. You'd need to add `coords.Distance`, `u.Magnitude`, `u.Dex`, `u.Decibel` to the list of `Quantity` subclasses that should be supported (where I'd need to check your code would work for the function quantities)

**Comment by taldcroft** [comment]
> 1. For SkyCoord, it turns out to be very handy that it has a list of things that are required to reproduce it faithfully. Perhaps we should have the same for other classes? E.g., in Time._apply, the list you have here is also used.

The basic strategy here was mostly looking at the class constructor API and then supplementing that with knowledge from methods like `Time._apply`.  But for instance with `SkyCoord` one doesn't get anything very useful by looking at `SkyCoord._apply` since it basically delegates to `__init__` with an input `SkyCoord` object.

Maybe your point here is that it would be handy if these objects had an `info.yaml_representer` method that returned a map (dict) of what is need to construct a copy (but see my response to your third point)?

> 2. Slightly relatedly, to what extent could we combine the information on how to serialise the objects with what pickle does? In the end, it needs the same information. Similarly, any effort to serialise in JSON would need the same information as well. Could we rely more on self.__dict__?

One could rely more on `self.__dict__`, but I think that is a slippery path to re-writing pickle.  Note that I believe yaml does use pickle for dumping arbitrary objects, but interestingly yaml fails to dump a Unit object:
```
>>> yaml.dump(u.m)
/Users/aldcroft/anaconda/envs/astropy3/lib/python3.3/copyreg.py in _reduce_ex(self, proto)
     63     else:
     64         if base is self.__class__:
---> 65             raise TypeError("can't pickle %s objects" % base.__name__)
     66         state = base(self)
     67     args = (self.__class__, base, state)

TypeError: can't pickle int objects
```
So the only way I could figure out how to accomplish YAML dump/load in a reasonable time was this hand-curated way.  But it is true that the representer functions could easily be adapted to JSON and that was something I had in mind.

> 3. There is something to be said for letting classes provide their own interface; this could perhaps be made part of info?

I thought about this as well (and even wrote code for that at one point).  The issue I have is that not all classes that we want to serialize will have an `info` attribute, and some like `np.ndarray` are not even under our control.  So at the end of the day it is probably just easier to have all the representers and constructors in one module?

**Comment by taldcroft** [comment]
BTW, I just remembered that this functionality requires the latest release of YAML (3.12) which fixes a problem where the dumper code is trying to compare ndarray objects to `None` or `()` and this makes Py3 unhappy.  (This is an odd story, because when running interactively that just generates a warning, but under pytest it turns into an exception.  I don't understand that.)

**Comment by MSeifert04** [comment]
> This is an odd story, because when running interactively that just generates a warning, but under pytest it turns into an exception. I don't understand that.

What kind of warning? For example deprecation warnings are automatically turned into exceptions in astropy-conftest.

**Comment by pllim** [comment]
Travis failed with `AssertionError` (files not closed) and `AttributeError`.

**Comment by mhvk** [comment]
@taldcroft - the problem with basic yaml not working on units is that `UnitBase` has a meta class `InheritDocstrings`. With our normal installation, even just
```
yaml.dump(u.UnitBase)
```
fails, while if I remove `@six.add_metaclass(InheritDocstrings)`, it works.

(This may well be a bug in `yaml`.)

**Comment by mhvk** [comment]
The problem is indeed the metaclass with yaml. E.g., in astropy 1.2.1, I can do
```
yaml.dump(Time)
# "!!python/name:astropy.time.core.Time ''\n"
```
but in current master it gives the same error as for `u.UnitBase` because `Time` now has the `ShapedLikeNDArray` metaclass.

**Comment by mhvk** [comment]
I don't know how relevant it is, but here is a work-around for the inability to serialize metaclasses:
```
import yaml
import astropy.units as u
class AstropyDumper(yaml.Dumper):
    pass

AstropyDumper.add_multi_representer(type, AstropyDumper.yaml_representers[type])
yaml.dump(u.m, Dumper=AstropyDumper)
```

(It seems python-yaml requires one to sign-up to submit bug reports; sadly, this means they won't get one...)

**Comment by taldcroft** [comment]
@mhvk - interesting and illuminating.  So will you OK in the end with a "hand-curated" approach which assumes that we can put a selected set of attributes into a flat dict and then use this in `__init__` (with some possible post-processing as in `Time`)?

As you can see the raw dump of `u.m` is a bit of a mess that I don't really want to dive into.  In other words, I'm not sure how to keep this idea both simple and safe (no arbitrary code) without using the curated approach.

File: astropy/utils/tests/test_yaml.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+except ImportError:
+    HAS_YAML = False
+
+pytestmark = pytest.mark.skipif('not HAS_YAML')
+
+
+@pytest.mark.parametrize('c', [u.m, u.m / u.s, u.hPa, u.dimensionless_unscaled])
+def test_unit(c):
+    cy = load(dump(c))
+    assert c == cy
```
Comment (mhvk) [inline astropy/utils/tests/test_yaml.py]: Actually, this should be true for all the units that are defined, i.e., all but `u.m / u.s`.

**Comment by mhvk** [comment]
@taldcroft - yes, good to get back to the higher-level question. I think it is fine to take a staged approach, i.e., start with what you have here, with stuff is hardcoded in the encoder/decoder, and try later to generalize it.

**Comment by eteq** [comment]
@taldcroft - this is very neat and a promising step!

Two large-scale design questions, though:
1. Why `*_constructor`/`*_representer` functions instead of `<class>._yaml_repr_`/`<class>._yaml_init_` or similar methods?  The latter strikes me as a good way to not have to manage the registry? (at least on the representer side, anyway)
2. This is fairly YAML-specific and meant to *stay* that way, right?  So the `load`/`dump` names should perhaps be `load_yaml`/`dump_yaml`?  (a bit nitpicking for a WIP, but it might be less work down the road to make such changes sooner)

**Comment by taldcroft** [comment]
> Why *_constructor/*_representer functions instead of <class>._yaml_repr_/<class>._yaml_init_ or similar methods? The latter strikes me as a good way to not have to manage the registry? (at least on the representer side, anyway)

There is at least one class (numpy) where we *can't* take that approach.  (Yes, upstream patch is possible but not really useful/practical here). 

But more to the point, having a registry gets a little tricky or unsavory when you think about details.  First, I think it would involve requiring some mixin metaclass for every YAML-enabled class in order to do the registry.  So I'm not super keen on that, but maybe people don't mind, e.g. putting this YAML metaclass on `Unit`?  

Then it gets slightly tricky with the `AstropyDumper/Loader` classes.  You see that right now all the supported classes get explicitly put into the class with class method calls when importing `utils.yaml`.  If the classes register themselves, I'm not sure it is possible to do this without importing `yaml` (the pyyaml version), and we definitely don't want to require importing `yaml` in order to import Quantity or Time.

So all in all I think the outside-in approach I've done is just conceptually simpler to develop, and probably maintain.  One can actually see all the constructors and representers in one place.

>This is fairly YAML-specific and meant to stay that way, right? So the load/dump names should perhaps be load_yaml/dump_yaml? (a bit nitpicking for a WIP, but it might be less work down the road to make such changes sooner)

About naming, I'm just following existing naming convention for the real `yaml`, e.g.:
```
from astropy.utils import yaml
...
out = yaml.dump(astropy_obj)
```
So effectively this `yaml` is exactly like the real `yaml` but it is astropy-enabled.

**Comment by taldcroft** [comment]
BTW much of the above was written thinking about *everything* happening automagically via a registry.  But taking a hybrid approach of defining `<class>._represent_as_dict()` and `<class>._construct_from_dict` which are essentially the existing functions, and then registering them manually in `utils/yaml.py` would be straightforward and probably make more sense.  (Except for numpy, of course).

That way those methods could in theory be used for JSON as well with just a little tweaking, and might also help out in making ECSV serialize mixin columns completely.

**Comment by mhvk** [comment]
On where to put things: above, I suggested similar that a class could provide the interface since then one doesn't have to keep a list. But at the same time, I don't think we want to make classes too cluttered with things that do not speak to their basic functionality (i.e., I think we should not have `Unit._to_yaml...`). This in part since for `Unit/Quantity` at least, I feel they should eventually go upstream (to scipy if not numpy). So, if we go the class-route, I'd propose to make this part of `info` (this not that strange -- it already helps to define how items are formatted within tables; one can start to see it as a general property where information relevant to other parts of astropy is stored).

In any case, I don't think this discussion has to hold up the PR: one can always move to a on-the-class scheme from the current outside-in approach (and, as @taldcroft notes, there will be special cases for objects like numpy arrays anyway).

**Comment by taldcroft** [comment]
@mhvk - made some progress, I hope you will like the changes.

**Comment by taldcroft** [comment]
This is ready for review now.  There is only one obvious issue now, namely in building the docs locally I get these warnings:
```
/Users/aldcroft/git/astropy/docs/api/astropy.io.misc.yaml.AstropyDumper.rst:7: WARNING: py:class reference target not found: yaml.dumper.SafeDumper
/Users/aldcroft/git/astropy/docs/api/astropy.io.misc.yaml.AstropyLoader.rst:7: WARNING: py:class reference target not found: yaml.loader.SafeLoader
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.representer.SafeRepresenter
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.composer.Composer
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.constructor.SafeConstructor
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.reader.Reader
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.emitter.Emitter
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.dumper.SafeDumper
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.parser.Parser
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.serializer.Serializer
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.scanner.Scanner
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.loader.SafeLoader
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.representer.BaseRepresenter
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.resolver.Resolver
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.constructor.BaseConstructor
/Users/aldcroft/git/astropy/docs/io/misc.rst:74: WARNING: py:class reference target not found: yaml.resolver.BaseResolver
```
It looks like in generating the API docs it wants to find definitions for various PyYaml objects.  These definitions don't exist anywhere AFAIK.  So I have no idea how to fix this.

**Comment by taldcroft** [comment]
The only other thing I can think of is a question about whether to inject `info` attributes into `Unit` and `EarthLocation`.  That would make all of the serialized object classes use `info` for the core represent / construct operations.  It makes no real difference for YAML stuff apart from introducing consistency.  I'm slightly concerned about unexpected side effects, however, esp. for `Unit`.

**Comment by taldcroft** [comment]
Doctest failures:
- Travis: unrelated timeouts and expected doc build failure mentioned above.  Actual code tests (py2.7 and py3.5 with optional dependencies) are passing.
- CircleCI: failure to build, presumably unrelated
- AppVeyor: float format difference in docstring, need `# doctest: +FLOAT_CMP`

**Comment by astrofrog** [comment]
@taldcroft - you need to rebase to get CircleCI to build (you are lacking the config)

**Comment by astrofrog** [comment]
@taldcroft - to fix the sphinx warnings, you either need to add pyyaml to the intersphinx modules, or whitelist those classes in the ``docs/nitpick-exceptions`` file.

File: astropy/io/misc/tests/test_yaml.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    for name in c.representation_component_names:
+        assert np.all(getattr(c, attr) == getattr(cy, attr))
+
+
+@pytest.mark.parametrize('frame', ['fk4', 'altaz'])
+def test_skycoord(frame):
+
+    c = SkyCoord([[1,2],[3,4]], [[5,6], [7,8]],
+                 unit='deg', frame=frame,
+                 obstime=Time.now(),
```
Comment (mhvk) [inline astropy/io/misc/tests/test_yaml.py]: I'd put a specific time here, so that we test only the serialisation.

File: astropy/io/misc/tests/test_yaml.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+except ImportError:
+    HAS_YAML = False
+
+pytestmark = pytest.mark.skipif('not HAS_YAML')
+
+
+@pytest.mark.parametrize('c', [u.m, u.m / u.s, u.hPa, u.dimensionless_unscaled])
+def test_unit(c):
+    cy = load(dump(c))
+    assert c == cy
```
Comment (mhvk) [inline astropy/io/misc/tests/test_yaml.py]: replace by
```
if isinstance(c, u.CompositeUnit):
    assert c == cy
else:
    assert c is cy
```

File: astropy/io/misc/tests/test_yaml.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    t.delta_ut1_utc = np.array([[3.0], [4.0]])
+    t.delta_tdb_tt = np.array([[5.0], [6.0]])
+    t.out_subfmt = 'date_hm'
+
+    return t
+
+
+def compare_time(t, ty):
+    assert type(t) is type(ty)
+    for attr in ('shape', 'jd1', 'jd2', 'format', 'scale', 'precision', 'in_subfmt',
```
Comment (mhvk) [inline astropy/io/misc/tests/test_yaml.py]: Maybe silly, so feel free to ignore, but somehow I also like the basic
```
assert np.all(t == ty)
```

File: astropy/io/misc/yaml.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+def _ndarray_representer(dumper, obj):
+    if obj.flags['C_CONTIGUOUS']:
+        obj_data = obj.data
+    else:
+        cont_obj = np.ascontiguousarray(obj)
+        assert(cont_obj.flags['C_CONTIGUOUS'])
+        obj_data = cont_obj.data
+    data_b64 = base64.b64encode(bytes(obj_data))
+    out = dict(__ndarray__=data_b64,
```
Comment (mhvk) [inline astropy/io/misc/yaml.py]: I think we should call this `buffer`, so that below we can read it in using the array constructor.

File: astropy/io/misc/yaml.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+               dtype=str(obj.dtype),
+               shape=obj.shape)
+
+    return dumper.represent_mapping(u'!numpy.ndarray', out)
+
+
+def _ndarray_constructor(loader, node):
+    map = loader.construct_mapping(node)
+    data = base64.b64decode(map['__ndarray__'])
+    return np.frombuffer(data, map['dtype']).reshape(map['shape'])
```
Comment (mhvk) [inline astropy/io/misc/yaml.py]: I *think* (i.e., I didn't test...) that one should just be able to do
```
np.ndarray(buffer=data, shape=map['shape'], dtype=map['dtype'])
```
It might also be an idea to mimic what `np.save`/`np.load` do: the only difference would be to also just use the data if they are `F_CONTIGUOUS` and, if so, add an `order='F'` to the mapping.

File: astropy/io/misc/yaml.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+
+
+def _earthlocation_representer(dumper, obj):
+    out = _get_obj_attrs_map(obj, ('x', 'y', 'z', 'ellipsoid'))
+    return dumper.represent_mapping(u'!astropy.coordinates.earth.EarthLocation', out)
+
+
+def _earthlocation_constructor(loader, node):
+    map = loader.construct_mapping(node)
+    ellipsoid = map.pop('ellipsoid')
```
Comment (mhvk) [inline astropy/io/misc/yaml.py]: Hmm, clearly two things are wrong with `EarthLocation`: it should have its own `info` (not just inherit from `Quantity`) and its `from_geocentric` class method should allow initializing `ellipsoid` -- but that can both be for a different PR...

Comment (mhvk) [inline astropy/io/misc/yaml.py]: Added this to the wishlist in #4261

File: astropy/io/misc/yaml.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+
+  >>> ty = yaml.load(td)
+  >>> ty
+  <Time object: scale='utc' format='mjd' value=2457389.0>
+
+  >>> ty.location
+  <EarthLocation (1000.0, 2000.0, 3000.0) km>
+"""
+
+from __future__ import absolute_import
```
Comment (mhvk) [inline astropy/io/misc/yaml.py]: Since you have multiple `u'` string prefixes, would it be an idea to add `unicode_literals` to the import here?  Also, I tend to always add `division`, just in case someone ever adds a division...

File: astropy/io/misc/yaml.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+    import yaml
+except ImportError:
+    raise ImportError('`import yaml` failed, PyYAML package is required for YAML')
+
+
+__all__ = ['AstropyLoader', 'AstropyDumper', 'load', 'load_all', 'dump']
+
+def _unit_representer(dumper, obj):
+    out = {'name': str(obj.to_string())}
+    return dumper.represent_mapping(u'!astropy.units.Unit', out)
```
Comment (mhvk) [inline astropy/io/misc/yaml.py]: See above for the `u'` -- I think we should write python3 code which happens to work on python2 -- so this should not be necessary...

File: astropy/units/quantity.py [hunk 10]
```diff
@@ -148,6 +149,12 @@ def possible_string_format_functions(format_):
         yield lambda format_, val: format_.format(val.value)
         yield lambda format_, val: format_ % val.value
 
+    def _construct_from_dict(self, map):
+        # Need to pop value because different Quantity subclasses use
+        # different first arg name for the value.  :-(
```
Comment (mhvk) [inline astropy/units/quantity.py]: Which is indeed not great...

File: astropy/time/core.py [hunk 11]
```diff
@@ -114,6 +117,40 @@ def unit(self):
...
+    _represent_as_dict_attrs = ('jd1', 'jd2', 'format', 'scale')
+
+    def _construct_from_dict(self, map):
+        format = map.pop('format')
+
+        map['format'] = 'jd'
+        map['val'] = map.pop('jd1')
+        map['val2'] = map.pop('jd2')
+
+        out = TimeDelta(**map)
```
Comment (mhvk) [inline astropy/time/core.py]: I think this should be `self._parent_cls(**map)`, no?

File: astropy/time/core.py [hunk 12]
```diff
@@ -114,6 +117,40 @@ def unit(self):
...
+    def _construct_from_dict(self, map):
+        format = map.pop('format')
+        delta_ut1_utc = map.pop('_delta_ut1_utc', None)
+        delta_tdb_tt = map.pop('_delta_tdb_tt', None)
+
+        map['format'] = 'jd'
+        map['val'] = map.pop('jd1')
+        map['val2'] = map.pop('jd2')
+
+        out = Time(**map)
```
Comment (mhvk) [inline astropy/time/core.py]: And here too, `self._parent_cls(**map)`?

**Comment by mhvk** [comment]
@taldcroft - I really like the approach with `info`; it comes into its own very well indeed! As you'll have seen, I had several comments, most very small, except perhaps to use the `ndarray` constructor instead of `frombuffer`. (Mostly as I like to get the naming in the `yaml` output right -- this would be the one thing that is hard to change after the fact, since we'd like yaml files to remain readable.)

As for your comments: I agree that `EarthLocation` should get its own `info` class (rather than just inheriting it from `Quantity`). For `Unit`, I feel the case is much less clear, since it is not array-like (and much of `info` assumes its parent class is). Overall, I think it is fine to treat it as a type of string (we may eventually have to add a `format` as well, but since right now this is not exposed, I don't see how to do that).

File: astropy/io/misc/yaml.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+try:
+    import yaml
+except ImportError:
+    raise ImportError('`import yaml` failed, PyYAML package is required for YAML')
+
+
+__all__ = ['AstropyLoader', 'AstropyDumper', 'load', 'load_all', 'dump']
+
+def _unit_representer(dumper, obj):
+    out = {'name': str(obj.to_string())}
```
Comment (mhvk) [inline astropy/io/misc/yaml.py]: Should we perhaps just call it `'unit'` instead of `'name'` -- the latter feels odd for something like `m/s^2`. (and, indeed, `CompositeUnit` instances do not have a `name` property).

**Comment by taldcroft** [comment]
@mhvk - thanks for the detailed review!  I'll be addressing these hopefully this evening.

File: astropy/io/misc/tests/test_yaml.py (hunk 2 above)
Comment (taldcroft) [inline astropy/io/misc/tests/test_yaml.py]: Done.

File: astropy/io/misc/tests/test_yaml.py (hunk 4 above)
Comment (taldcroft) [inline astropy/io/misc/tests/test_yaml.py]: Done.

File: astropy/io/misc/yaml.py (hunk 7 above)
Comment (taldcroft) [inline astropy/io/misc/yaml.py]: I had not realized that `EarthLocation` is a `Quantity` subclass.  Giving it its own `info` in 3f778c1 allowed making it just another one of the list of `Quantity` subclasses, nice!

File: astropy/io/misc/yaml.py (hunk 8 above)
Comment (taldcroft) [inline astropy/io/misc/yaml.py]: I took away all the `u'` string prefixes.  I never specifically intended to use unicode, I think that just crept in at some point from copy/paste.

File: astropy/io/misc/yaml.py (hunk 9 above)
Comment (taldcroft) [inline astropy/io/misc/yaml.py]: Done.

File: astropy/time/core.py (hunk 11 above)
Comment (taldcroft) [inline astropy/time/core.py]: Done.

File: astropy/time/core.py (hunk 12 above)
Comment (taldcroft) [inline astropy/time/core.py]: Done.

**Comment by taldcroft** [comment]
@mhvk - I think I have addressed all comments.  Cross fingers that tests and doc build pass.

**Comment by taldcroft** [comment]
@mhvk - the test failure is related to the following funny that appears only in Python 3.  This makes **no** sense to me and clearly I don't understand something.  On Python 2 the byte order is changed in the way that I would expect.
```
In [13]: f = bytes(np.array([[0,1], [2,3]], order='F').data)

In [14]: f
Out[14]: b'\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00'

In [15]: c = bytes(np.array([[0,1], [2,3]], order='C').data)

In [16]: c
Out[16]: b'\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00'

In [17]: f == c
Out[17]: True
```

**Comment by taldcroft** [comment]
@mhvk - I found a workaround though I still don't understand the original problem.

**Comment by mhvk** [comment]
A really nice addition!!
